File: src/bfs-field.js
function toKey(location) {
    if (typeof location === 'string') {
        return location
    }

    if (Number.isFinite(location?.x) && Number.isFinite(location?.y)) {
        return location.x + "," + location.y
    }

    return null
}

function toCoords(coords) {
    if (coords && typeof coords === 'object' && Number.isFinite(coords.x) && Number.isFinite(coords.y)) {
        return { x: coords.x, y: coords.y }
    }

    try {
        const [x, y] = String(coords).split(",").map(Number)
        return { x, y }
    } catch (error) {
        return coords
    }
}

function normalizeTileKey(tile) {
    if (tile === null || tile === undefined) {
        return null
    }

    if (typeof tile === 'string') {
        return tile
    }

    if (typeof tile === 'object') {
        if (typeof tile.key === 'string') {
            return tile.key
        }

        if (Number.isFinite(tile.x) && Number.isFinite(tile.y)) {
            return toKey(tile)
        }
    }

    return null
}

export class BfsField {

    constructor({
        name = '',
        type = 'shared',
        cave = null,
        ownerBuilding = null
    } = {}) {
        this.name = typeof name === 'string' ? name : ''
        this.type = typeof type === 'string' ? type : 'shared'
        this.cave = cave
        this.ownerBuilding = ownerBuilding
        this.field = new Map()
        this.updated = false
        this.updatedTiles = new Set()
        this.updatedBuildings = new Set()
        this.updatedCreatures = new Set()
        this.trackedBuildings = new Set()
        this.trackedCreatures = new Set()
    }

    getName() {
        return this.name
    }

    getType() {
        return this.type
    }

    setCave(cave) {
        this.cave = cave
        return this.cave
    }

    getCave() {
        return this.cave
    }

    setOwnerBuilding(building) {
        this.ownerBuilding = building
        return this.ownerBuilding
    }

    getOwnerBuilding() {
        return this.ownerBuilding
    }

    setField(field) {
        this.field = field instanceof Map ? field : new Map()
        return this.field
    }

    commitField(field) {
        this.setField(field)
        this.clearUpdates()
        return this.field
    }

    getField({ refresh = true } = {}) {
        return refresh ? this.refresh() : this.field
    }

    isUpdated() {
        return this.updated === true
    }

    getUpdatedTileKeys() {
        return [...this.updatedTiles]
    }

    getUpdatedBuildings() {
        return [...this.updatedBuildings]
    }

    getUpdatedCreatures() {
        return [...this.updatedCreatures]
    }

    getTrackedBuildings() {
        return [...this.trackedBuildings]
    }

    getTrackedCreatures() {
        return [...this.trackedCreatures]
    }

    setTrackedTargets({ buildings = [], creatures = [] } = {}) {
        this.trackedBuildings.clear()
        this.trackedCreatures.clear()

        for (const building of buildings) {
            if (building) {
                this.trackedBuildings.add(building)
            }
        }

        for (const creature of creatures) {
            if (creature) {
                this.trackedCreatures.add(creature)
            }
        }

        return {
            buildings: this.getTrackedBuildings(),
            creatures: this.getTrackedCreatures()
        }
    }

    clearUpdates() {
        this.updated = true
        this.updatedTiles.clear()
        this.updatedBuildings.clear()
        this.updatedCreatures.clear()
        return this.updated
    }

    markTilesDirty(tileKeys = []) {
        this.updated = false

        if (!Array.isArray(tileKeys)) {
            return this.updated
        }

        for (const tileKey of tileKeys) {
            const normalizedTileKey = normalizeTileKey(tileKey)
            if (normalizedTileKey !== null) {
                this.updatedTiles.add(normalizedTileKey)
            }
        }

        return this.updated
    }

    markBuildingsDirty(buildings = []) {
        this.updated = false

        if (!Array.isArray(buildings)) {
            return this.updated
        }

        for (const building of buildings) {
            if (building) {
                this.updatedBuildings.add(building)
            }
        }

        return this.updated
    }

    markCreaturesDirty(creatures = []) {
        this.updated = false

        if (!Array.isArray(creatures)) {
            return this.updated
        }

        for (const creature of creatures) {
            if (creature) {
                this.updatedCreatures.add(creature)
            }
        }

        return this.updated
    }

    markDirty({ tileKeys = [], buildings = [], creatures = [] } = {}) {
        this.updated = false
        this.markTilesDirty(tileKeys)
        this.markBuildingsDirty(buildings)
        this.markCreaturesDirty(creatures)
        return this.updated
    }

    hasActiveBuildingTarget() {
        if (this.type !== 'building') {
            return false
        }

        const building = this.getOwnerBuilding()
        return Boolean(
            this.cave &&
            building &&
            Array.isArray(building.tileArray) &&
            building.tileArray.length > 0
        )
    }

    getTile(tileOrKey) {
        const tileKey = normalizeTileKey(tileOrKey)
        if (tileKey === null) {
            return null
        }

        if (typeof this.cave?.getTile !== 'function') {
            return null
        }

        return this.cave.getTile(tileKey) ?? null
    }

    isTileInCoverage(tileOrKey) {
        const tile = this.getTile(tileOrKey)
        if (!tile) {
            return false
        }

        if (this.type === 'building') {
            return this.hasActiveBuildingTarget() && this.cave?.isTileReachable?.(tile) === true
        }

        return this.cave?.isTileRevealed?.(tile) === true
    }

    getCoverageTiles() {
        if (!this.cave) {
            return []
        }

        if (this.type === 'building') {
            if (!this.hasActiveBuildingTarget()) {
                return []
            }

            return typeof this.cave.getReachableTiles === 'function'
                ? this.cave.getReachableTiles()
                : []
        }

        if (typeof this.cave.getTiles !== 'function') {
            return []
        }

        return this.cave.getTiles().filter((tile) => this.isTileInCoverage(tile))
    }

    createBaseField() {
        const field = new Map()

        for (const tile of this.getCoverageTiles()) {
            field.set(tile.key, Infinity)
        }

        return field
    }

    syncCoverage(field = this.field) {
        if (!(field instanceof Map)) {
            return null
        }

        for (const tileKey of [...field.keys()]) {
            if (!this.isTileInCoverage(tileKey)) {
                field.delete(tileKey)
            }
        }

        for (const tile of this.getCoverageTiles()) {
            if (!field.has(tile.key)) {
                field.set(tile.key, Infinity)
            }
        }

        return field
    }

    addAdjacentPassableSeeds(tile, blockedKeys, seedKeys) {
        if (!tile) {
            return
        }

        for (const neighbor of tile.getNeighbors()) {
            if (!this.isTileInCoverage(neighbor) || !neighbor?.creatureFits() || blockedKeys.has(neighbor.key)) {
                continue
            }

            seedKeys.add(neighbor.key)
        }
    }

    addBuildingTargets(building, blockedKeys, seedKeys, { blockPassableTiles = false } = {}) {
        if (!building || !Array.isArray(building.tileArray) || building.tileArray.length === 0) {
            return
        }

        for (const tile of building.tileArray) {
            if (!tile) {
                continue
            }

            const shouldBlockTile = blockPassableTiles || !tile.creatureFits()
            if (!shouldBlockTile) {
                continue
            }

            blockedKeys.add(tile.key)
            this.addAdjacentPassableSeeds(tile, blockedKeys, seedKeys)
        }
    }

    buildBuildingSeedKeys(building) {
        const seedKeys = new Set()
        if (!building || !Array.isArray(building.tileArray) || building.tileArray.length === 0) {
            return seedKeys
        }

        for (const tile of building.tileArray) {
            if (!tile?.creatureFits() || !this.isTileInCoverage(tile)) {
                continue
            }

            seedKeys.add(tile.key)
        }

        if (seedKeys.size > 0) {
            return seedKeys
        }

        for (const tile of building.tileArray) {
            if (!tile) {
                continue
            }

            for (const neighbor of tile.getNeighbors()) {
                if (!neighbor?.creatureFits() || !this.isTileInCoverage(neighbor)) {
                    continue
                }

                seedKeys.add(neighbor.key)
            }
        }

        return seedKeys
    }

    buildSnapshot() {
        const blockedKeys = new Set()
        const seedKeys = new Set()
        const trackedBuildings = []
        const trackedCreatures = []

        if (!this.cave) {
            this.setTrackedTargets()
            return {
                blockedKeys,
                seedKeys
            }
        }

        if (typeof this.cave.getTiles === 'function') {
            for (const tile of this.cave.getTiles()) {
                if (!this.isTileInCoverage(tile) || !tile?.creatureFits()) {
                    blockedKeys.add(tile.key)
                }
            }
        }

        if (this.type === 'building') {
            const building = this.getOwnerBuilding()
            if (this.hasActiveBuildingTarget()) {
                trackedBuildings.push(building)

                for (const seedKey of this.buildBuildingSeedKeys(building)) {
                    seedKeys.add(seedKey)
                }
            }
        } else if (this.type === 'enemy') {
            for (const creature of this.cave.enemies ?? []) {
                trackedCreatures.push(creature)
                const tile = this.getTile(toKey(creature.location))
                if (!tile) {
                    continue
                }

                blockedKeys.add(tile.key)
                this.addAdjacentPassableSeeds(tile, blockedKeys, seedKeys)
            }
        } else if (this.type === 'colony') {
            for (const creature of this.cave.trilobites ?? []) {
                trackedCreatures.push(creature)
                const tile = this.getTile(toKey(creature.location))
                if (!tile) {
                    continue
                }

                blockedKeys.add(tile.key)
                this.addAdjacentPassableSeeds(tile, blockedKeys, seedKeys)
            }

            for (const building of this.cave.buildings ?? []) {
                if (!building) {
                    continue
                }

                trackedBuildings.push(building)
                const isAlgaeFarm = building?.name === 'Algae Farm' || building?.constructor?.name === 'AlgaeFarm'
                this.addBuildingTargets(building, blockedKeys, seedKeys, {
                    blockPassableTiles: isAlgaeFarm
                })
            }
        }

        this.setTrackedTargets({
            buildings: trackedBuildings,
            creatures: trackedCreatures
        })

        return {
            blockedKeys,
            seedKeys
        }
    }

    computeValue(tileKey, field, snapshot) {
        const tile = this.getTile(tileKey)
        if (!tile || !this.isTileInCoverage(tile) || snapshot.blockedKeys.has(tileKey)) {
            return Infinity
        }

        if (snapshot.seedKeys.has(tileKey)) {
            return 0
        }

        let bestNeighborValue = Infinity

        for (const neighbor of tile.getNeighbors()) {
            if (!this.isTileInCoverage(neighbor) || snapshot.blockedKeys.has(neighbor.key)) {
                continue
            }

            const neighborValue = field.get(neighbor.key) ?? Infinity
            if (neighborValue < bestNeighborValue) {
                bestNeighborValue = neighborValue
            }
        }

        if (!Number.isFinite(bestNeighborValue)) {
            return Infinity
        }

        return bestNeighborValue + 1
    }

    rebuild() {
        const snapshot = this.buildSnapshot()
        const field = this.createBaseField()
        const queue = []
        let queueHead = 0

        for (const seedKey of snapshot.seedKeys) {
            if (snapshot.blockedKeys.has(seedKey) || !field.has(seedKey)) {
                continue
            }

            field.set(seedKey, 0)
            queue.push(seedKey)
        }

        while (queueHead < queue.length) {
            const currentKey = queue[queueHead]
            queueHead++

            const currentTile = this.getTile(currentKey)
            if (!currentTile) {
                continue
            }

            const currentValue = field.get(currentKey) ?? Infinity
            if (!Number.isFinite(currentValue)) {
                continue
            }

            for (const neighbor of currentTile.getNeighbors()) {
                if (!this.isTileInCoverage(neighbor) || snapshot.blockedKeys.has(neighbor.key)) {
                    continue
                }

                const nextValue = currentValue + 1
                if (nextValue >= (field.get(neighbor.key) ?? Infinity)) {
                    continue
                }

                field.set(neighbor.key, nextValue)
                queue.push(neighbor.key)
            }
        }

        return this.commitField(field)
    }

    rebalance(dirtyKeys = this.getUpdatedTileKeys()) {
        if (!(this.field instanceof Map) || this.field.size === 0 || !Array.isArray(dirtyKeys) || dirtyKeys.length === 0) {
            return this.rebuild()
        }

        const field = this.syncCoverage(this.field)
        const snapshot = this.buildSnapshot()
        const queue = []
        const queued = new Set()
        let queueHead = 0

        const enqueue = (tileKey) => {
            const normalizedTileKey = normalizeTileKey(tileKey)
            if (normalizedTileKey === null || queued.has(normalizedTileKey)) {
                return
            }

            const tile = this.getTile(normalizedTileKey)
            if (!tile || !this.isTileInCoverage(tile)) {
                return
            }

            queued.add(normalizedTileKey)
            queue.push(normalizedTileKey)
        }

        for (const dirtyKey of dirtyKeys) {
            const tile = this.getTile(dirtyKey)
            if (!tile) {
                continue
            }

            enqueue(tile.key)
            for (const neighbor of tile.getNeighbors()) {
                enqueue(neighbor.key)
            }
        }

        while (queueHead < queue.length) {
            const currentKey = queue[queueHead]
            queueHead++
            queued.delete(currentKey)

            const currentValue = field.get(currentKey) ?? Infinity
            const nextValue = this.computeValue(currentKey, field, snapshot)
            if (Object.is(currentValue, nextValue)) {
                continue
            }

            field.set(currentKey, nextValue)

            const currentTile = this.getTile(currentKey)
            if (!currentTile) {
                continue
            }

            for (const neighbor of currentTile.getNeighbors()) {
                enqueue(neighbor.key)
            }
        }

        return this.commitField(field)
    }

    refresh() {
        if (!(this.field instanceof Map) || this.field.size === 0) {
            return this.rebuild()
        }

        if (this.isUpdated()) {
            return this.field
        }

        const dirtyKeys = this.getUpdatedTileKeys()
        if (dirtyKeys.length === 0) {
            return this.rebuild()
        }

        return this.rebalance(dirtyKeys)
    }

    getFieldValue(location, { refresh = true } = {}) {
        const field = this.getField({ refresh })
        if (!(field instanceof Map) || !location) {
            return Infinity
        }

        const key = toKey(location)
        return key === null ? Infinity : (field.get(key) ?? Infinity)
    }

    getNextStep(location, { refresh = true } = {}) {
        const field = this.getField({ refresh })
        if (!(field instanceof Map) || !location) {
            return null
        }

        const currentKey = toKey(location)
        const currentTile = this.getTile(currentKey)
        if (!currentTile) {
            return null
        }

        const currentValue = field.get(currentKey) ?? Infinity
        let bestNeighbor = null
        let bestValue = currentValue

        for (const neighbor of currentTile.getNeighbors()) {
            if (!neighbor.creatureFits()) {
                continue
            }

            const neighborValue = field.get(neighbor.key) ?? Infinity
            if (!Number.isFinite(neighborValue) || neighborValue >= bestValue) {
                continue
            }

            if (bestNeighbor === null || neighborValue < bestValue || (neighborValue === bestValue && neighbor.key < bestNeighbor.key)) {
                bestNeighbor = neighbor
                bestValue = neighborValue
            }
        }

        return bestNeighbor ? toCoords(bestNeighbor.key) : null
    }

    buildPathFrom(startLocation, { refresh = true } = {}) {
        const field = this.getField({ refresh })
        if (!(field instanceof Map) || !startLocation) {
            return null
        }

        const startKey = toKey(startLocation)
        const startValue = field.get(startKey) ?? Infinity
        if (!Number.isFinite(startValue)) {
            return null
        }

        const path = [{ x: startLocation.x, y: startLocation.y, type: 'move' }]
        let currentLocation = { x: startLocation.x, y: startLocation.y }
        let currentValue = startValue
        let timeCount = 0

        while (currentValue > 0 && timeCount < 7850) {
            const nextLocation = this.getNextStep(currentLocation, { refresh: false })
            if (!nextLocation) {
                return null
            }

            path.push({ x: nextLocation.x, y: nextLocation.y, type: 'move' })
            currentLocation = nextLocation
            currentValue = field.get(toKey(currentLocation)) ?? Infinity
            timeCount++
        }

        if (currentValue !== 0) {
            return null
        }

        return path
    }

}

File: TriloGame.CSharp/src/TriloGame.Game/Core/Pathfinding/BfsField.cs
using TriloGame.Game.Core.Buildings;
using TriloGame.Game.Core.Entities;
using TriloGame.Game.Core.World;
using TriloGame.Game.Shared.Math;

namespace TriloGame.Game.Core.Pathfinding;

public sealed class BfsField
{
    private int[] _values = [];
    private bool[] _covered = [];
    private bool[] _nextCovered = [];
    private bool[] _blocked = [];
    private bool[] _seeded = [];
    private bool[] _queued = [];
    private readonly Queue<int> _queue = [];
    private readonly List<int> _seedIds = [];
    private bool _fieldCacheDirty = true;
    private int _coverageCount;

    public BfsField(string name = "", string type = "shared", Cave? cave = null, Building? ownerBuilding = null)
    {
        Name = name;
        Type = type;
        Cave = cave;
        OwnerBuilding = ownerBuilding;
        Field = new Dictionary<string, int>(StringComparer.Ordinal);
        UpdatedTiles = new HashSet<string>(StringComparer.Ordinal);
        UpdatedBuildings = [];
        UpdatedCreatures = [];
        TrackedBuildings = [];
        TrackedCreatures = [];
        EnsureCapacity(cave?.TileCapacity ?? 0);
    }

    public string Name { get; }

    public string Type { get; }

    public Cave? Cave { get; private set; }

    public Building? OwnerBuilding { get; private set; }

    public Dictionary<string, int> Field { get; private set; }

    public bool Updated { get; private set; }

    public HashSet<string> UpdatedTiles { get; }

    public HashSet<Building> UpdatedBuildings { get; }

    public HashSet<Creature> UpdatedCreatures { get; }

    public HashSet<Building> TrackedBuildings { get; }

    public HashSet<Creature> TrackedCreatures { get; }

    public void SetCave(Cave? cave)
    {
        Cave = cave;
        EnsureCapacity(cave?.TileCapacity ?? 0);
        _fieldCacheDirty = true;
    }

    public void SetOwnerBuilding(Building? building)
    {
        OwnerBuilding = building;
    }

    public void SetField(Dictionary<string, int>? field)
    {
        Field = field is null
            ? new Dictionary<string, int>(StringComparer.Ordinal)
            : new Dictionary<string, int>(field, StringComparer.Ordinal);
        _fieldCacheDirty = false;
        ImportField(Field);
    }

    public Dictionary<string, int> CommitField(Dictionary<string, int> field)
    {
        SetField(field);
        ClearUpdates();
        return Field;
    }

    public bool IsUpdated() => Updated;

    public void SetTrackedTargets(IEnumerable<Building>? buildings = null, IEnumerable<Creature>? creatures = null)
    {
        TrackedBuildings.Clear();
        TrackedCreatures.Clear();

        foreach (var building in buildings ?? [])
        {
            TrackedBuildings.Add(building);
        }

        foreach (var creature in creatures ?? [])
        {
            TrackedCreatures.Add(creature);
        }
    }

    public bool ClearUpdates()
    {
        Updated = true;
        UpdatedTiles.Clear();
        UpdatedBuildings.Clear();
        UpdatedCreatures.Clear();
        return Updated;
    }

    public bool MarkTilesDirty(IEnumerable<string>? tileKeys)
    {
        Updated = false;
        foreach (var tileKey in tileKeys ?? [])
        {
            if (!string.IsNullOrWhiteSpace(tileKey))
            {
                UpdatedTiles.Add(tileKey);
            }
        }

        return Updated;
    }

    public bool MarkBuildingsDirty(IEnumerable<Building>? buildings)
    {
        Updated = false;
        foreach (var building in buildings ?? [])
        {
            UpdatedBuildings.Add(building);
        }

        return Updated;
    }

    public bool MarkCreaturesDirty(IEnumerable<Creature>? creatures)
    {
        Updated = false;
        foreach (var creature in creatures ?? [])
        {
            UpdatedCreatures.Add(creature);
        }

        return Updated;
    }

    public bool MarkDirty(IEnumerable<string>? tileKeys, IEnumerable<Building>? buildings, IEnumerable<Creature>? creatures)
    {
        Updated = false;
        MarkTilesDirty(tileKeys);
        MarkBuildingsDirty(buildings);
        MarkCreaturesDirty(creatures);
        return Updated;
    }

    public Dictionary<string, int> GetField(bool refresh = true)
    {
        if (refresh)
        {
            Refresh();
        }

        if (!_fieldCacheDirty)
        {
            return Field;
        }

        var field = new Dictionary<string, int>(Math.Max(0, _coverageCount), StringComparer.Ordinal);
        if (Cave is not null)
        {
            foreach (var tile in Cave.GetTiles())
            {
                if (_covered[tile.Id])
                {
                    field[tile.Key] = _values[tile.Id];
                }
            }
        }

        Field = field;
        _fieldCacheDirty = false;
        return Field;
    }

    public bool HasActiveBuildingTarget()
    {
        return string.Equals(Type, "building", StringComparison.Ordinal) &&
               Cave is not null &&
               OwnerBuilding is not null &&
               OwnerBuilding.TileArray.Count > 0;
    }

    public Tile? GetTile(string? tileKey)
    {
        if (string.IsNullOrWhiteSpace(tileKey) || Cave is null)
        {
            return null;
        }

        return Cave.GetTile(tileKey);
    }

    private void EnsureCapacity(int requiredCapacity)
    {
        if (requiredCapacity <= _values.Length)
        {
            return;
        }

        var oldLength = _values.Length;
        var newLength = Math.Max(requiredCapacity, Math.Max(8, oldLength * 2));

        Array.Resize(ref _values, newLength);
        Array.Fill(_values, int.MaxValue, oldLength, newLength - oldLength);
        Array.Resize(ref _covered, newLength);
        Array.Resize(ref _nextCovered, newLength);
        Array.Resize(ref _blocked, newLength);
        Array.Resize(ref _seeded, newLength);
        Array.Resize(ref _queued, newLength);
    }

    private void ImportField(Dictionary<string, int> field)
    {
        if (Cave is null)
        {
            _coverageCount = 0;
            return;
        }

        EnsureCapacity(Cave.TileCapacity);
        Array.Clear(_covered, 0, _covered.Length);
        _coverageCount = 0;

        foreach (var tile in Cave.GetTiles())
        {
            _values[tile.Id] = int.MaxValue;
        }

        foreach (var pair in field)
        {
            var tile = Cave.GetTile(pair.Key);
            if (tile is null)
            {
                continue;
            }

            if (!_covered[tile.Id])
            {
                _coverageCount++;
            }

            _covered[tile.Id] = true;
            _values[tile.Id] = pair.Value;
        }
    }

    private bool IsTileInCoverage(Tile? tile)
    {
        if (tile is null || Cave is null)
        {
            return false;
        }

        if (string.Equals(Type, "building", StringComparison.Ordinal))
        {
            return HasActiveBuildingTarget() && Cave.IsTileReachable(tile);
        }

        return Cave.IsTileRevealed(tile);
    }

    private void RefreshCoverageState(bool resetValues)
    {
        if (Cave is null)
        {
            _coverageCount = 0;
            Array.Clear(_covered, 0, _covered.Length);
            return;
        }

        EnsureCapacity(Cave.TileCapacity);
        Array.Clear(_nextCovered, 0, _nextCovered.Length);

        if (string.Equals(Type, "building", StringComparison.Ordinal))
        {
            if (HasActiveBuildingTarget())
            {
                foreach (var tile in Cave.GetReachableTiles())
                {
                    _nextCovered[tile.Id] = true;
                }
            }
        }
        else
        {
            foreach (var tile in Cave.GetTiles())
            {
                if (IsTileInCoverage(tile))
                {
                    _nextCovered[tile.Id] = true;
                }
            }
        }

        _coverageCount = 0;
        foreach (var tile in Cave.GetTiles())
        {
            var tileId = tile.Id;
            var shouldCover = _nextCovered[tileId];
            _covered[tileId] = shouldCover;
            if (shouldCover)
            {
                _coverageCount++;
                if (resetValues)
                {
                    _values[tileId] = int.MaxValue;
                }
            }
            else
            {
                _values[tileId] = int.MaxValue;
            }
        }

        _fieldCacheDirty = true;
    }

    private void AddSeed(Tile tile)
    {
        if (_seeded[tile.Id])
        {
            return;
        }

        _seeded[tile.Id] = true;
        _seedIds.Add(tile.Id);
    }

    private void AddAdjacentPassableSeeds(Tile? tile)
    {
        if (tile is null)
        {
            return;
        }

        foreach (var neighbor in tile.Neighbors)
        {
            if (!_covered[neighbor.Id] || !neighbor.CreatureFits() || _blocked[neighbor.Id])
            {
                continue;
            }

            AddSeed(neighbor);
        }
    }

    private void AddBuildingTargets(Building? building, bool blockPassableTiles = false)
    {
        if (building is null || building.TileArray.Count == 0)
        {
            return;
        }

        foreach (var tile in building.TileArray)
        {
            var shouldBlockTile = blockPassableTiles || !tile.CreatureFits();
            if (!shouldBlockTile)
            {
                continue;
            }

            _blocked[tile.Id] = true;
            AddAdjacentPassableSeeds(tile);
        }
    }

    private void AddBuildingSeedIds(Building? building)
    {
        if (building is null || building.TileArray.Count == 0)
        {
            return;
        }

        foreach (var tile in building.TileArray)
        {
            if (tile.CreatureFits() && _covered[tile.Id])
            {
                AddSeed(tile);
            }
        }

        if (_seedIds.Count > 0)
        {
            return;
        }

        foreach (var tile in building.TileArray)
        {
            foreach (var neighbor in tile.Neighbors)
            {
                if (neighbor.CreatureFits() && _covered[neighbor.Id])
                {
                    AddSeed(neighbor);
                }
            }
        }
    }

    private void BuildSnapshot()
    {
        var trackedBuildings = new List<Building>();
        var trackedCreatures = new List<Creature>();

        Array.Clear(_blocked, 0, _blocked.Length);
        Array.Clear(_seeded, 0, _seeded.Length);
        _seedIds.Clear();

        if (Cave is null)
        {
            SetTrackedTargets();
            return;
        }

        foreach (var tile in Cave.GetTiles())
        {
            if (!_covered[tile.Id] || !tile.CreatureFits())
            {
                _blocked[tile.Id] = true;
            }
        }

        if (string.Equals(Type, "building", StringComparison.Ordinal))
        {
            if (HasActiveBuildingTarget())
            {
                trackedBuildings.Add(OwnerBuilding!);
                AddBuildingSeedIds(OwnerBuilding);
            }
        }
        else if (string.Equals(Type, "enemy", StringComparison.Ordinal))
        {
            foreach (var creature in Cave.GetEnemyList())
            {
                trackedCreatures.Add(creature);
                var tile = GetTile(creature.Location.ToString());
                if (tile is null)
                {
                    continue;
                }

                _blocked[tile.Id] = true;
                AddAdjacentPassableSeeds(tile);
            }
        }
        else if (string.Equals(Type, "colony", StringComparison.Ordinal))
        {
            foreach (var creature in Cave.GetTrilobiteList())
            {
                trackedCreatures.Add(creature);
                var tile = GetTile(creature.Location.ToString());
                if (tile is null)
                {
                    continue;
                }

                _blocked[tile.Id] = true;
                AddAdjacentPassableSeeds(tile);
            }

            foreach (var building in Cave.GetBuildingList())
            {
                trackedBuildings.Add(building);
                var isAlgaeFarm = string.Equals(building.Name, "Algae Farm", StringComparison.Ordinal) ||
                                  string.Equals(building.GetType().Name, "AlgaeFarm", StringComparison.Ordinal);
                AddBuildingTargets(building, isAlgaeFarm);
            }
        }

        SetTrackedTargets(trackedBuildings, trackedCreatures);
    }

    private int ComputeValue(Tile tile)
    {
        if (!_covered[tile.Id] || _blocked[tile.Id])
        {
            return int.MaxValue;
        }

        if (_seeded[tile.Id])
        {
            return 0;
        }

        var bestNeighbor = int.MaxValue;
        foreach (var neighbor in tile.Neighbors)
        {
            if (!_covered[neighbor.Id] || _blocked[neighbor.Id])
            {
                continue;
            }

            var neighborValue = _values[neighbor.Id];
            if (neighborValue < bestNeighbor)
            {
                bestNeighbor = neighborValue;
            }
        }

        return bestNeighbor == int.MaxValue ? int.MaxValue : bestNeighbor + 1;
    }

    private void EnqueueTile(Tile tile)
    {
        if (!_covered[tile.Id] || _queued[tile.Id])
        {
            return;
        }

        _queued[tile.Id] = true;
        _queue.Enqueue(tile.Id);
    }

    private Dictionary<string, int> CommitCurrentField()
    {
        _fieldCacheDirty = true;
        ClearUpdates();
        return GetField(false);
    }

    public Dictionary<string, int> Rebuild()
    {
        if (Cave is null)
        {
            Field = new Dictionary<string, int>(StringComparer.Ordinal);
            _fieldCacheDirty = false;
            ClearUpdates();
            return Field;
        }

        RefreshCoverageState(resetValues: true);
        BuildSnapshot();
        Array.Clear(_queued, 0, _queued.Length);
        _queue.Clear();

        foreach (var seedId in _seedIds)
        {
            if (!_covered[seedId] || _blocked[seedId])
            {
                continue;
            }

            _values[seedId] = 0;
            if (!_queued[seedId])
            {
                _queued[seedId] = true;
                _queue.Enqueue(seedId);
            }
        }

        while (_queue.Count > 0)
        {
            var currentId = _queue.Dequeue();
            _queued[currentId] = false;

            var currentTile = Cave.GetTileById(currentId);
            if (currentTile is null)
            {
                continue;
            }

            var currentValue = _values[currentId];
            if (currentValue == int.MaxValue)
            {
                continue;
            }

            foreach (var neighbor in currentTile.Neighbors)
            {
                if (!_covered[neighbor.Id] || _blocked[neighbor.Id])
                {
                    continue;
                }

                var nextValue = currentValue + 1;
                if (nextValue >= _values[neighbor.Id])
                {
                    continue;
                }

                _values[neighbor.Id] = nextValue;
                EnqueueTile(neighbor);
            }
        }

        return CommitCurrentField();
    }

    public Dictionary<string, int> Rebalance(IEnumerable<string>? dirtyKeys = null)
    {
        if (Cave is null || _coverageCount == 0)
        {
            return Rebuild();
        }

        RefreshCoverageState(resetValues: false);
        BuildSnapshot();
        Array.Clear(_queued, 0, _queued.Length);
        _queue.Clear();

        var hasDirty = false;
        foreach (var dirtyKey in dirtyKeys ?? UpdatedTiles)
        {
            var tile = GetTile(dirtyKey);
            if (tile is null)
            {
                continue;
            }

            hasDirty = true;
            EnqueueTile(tile);
            foreach (var neighbor in tile.Neighbors)
            {
                EnqueueTile(neighbor);
            }
        }

        if (!hasDirty)
        {
            return Rebuild();
        }

        while (_queue.Count > 0)
        {
            var currentId = _queue.Dequeue();
            _queued[currentId] = false;

            var currentTile = Cave.GetTileById(currentId);
            if (currentTile is null)
            {
                continue;
            }

            var nextValue = ComputeValue(currentTile);
            if (_values[currentId] == nextValue)
            {
                continue;
            }

            _values[currentId] = nextValue;
            foreach (var neighbor in currentTile.Neighbors)
            {
                EnqueueTile(neighbor);
            }
        }

        return CommitCurrentField();
    }

    public Dictionary<string, int> Refresh()
    {
        if (_coverageCount == 0)
        {
            return Rebuild();
        }

        if (IsUpdated())
        {
            return GetField(false);
        }

        return UpdatedTiles.Count == 0 ? Rebuild() : Rebalance();
    }

    public int GetFieldValue(GridPoint location, bool refresh = true)
    {
        if (refresh)
        {
            Refresh();
        }

        var tile = GetTile(location.ToString());
        return tile is null || !_covered[tile.Id]
            ? int.MaxValue
            : _values[tile.Id];
    }

    public GridPoint? GetNextStep(GridPoint location, bool refresh = true)
    {
        if (refresh)
        {
            Refresh();
        }

        var currentTile = GetTile(location.ToString());
        if (currentTile is null || !_covered[currentTile.Id])
        {
            return null;
        }

        var currentValue = _values[currentTile.Id];
        Tile? bestNeighbor = null;
        var bestValue = currentValue;

        foreach (var neighbor in currentTile.Neighbors)
        {
            if (!neighbor.CreatureFits() || !_covered[neighbor.Id])
            {
                continue;
            }

            var neighborValue = _values[neighbor.Id];
            if (neighborValue == int.MaxValue || neighborValue >= bestValue)
            {
                continue;
            }

            if (bestNeighbor is null || neighborValue < bestValue || string.CompareOrdinal(neighbor.Key, bestNeighbor.Key) < 0)
            {
                bestNeighbor = neighbor;
                bestValue = neighborValue;
            }
        }

        return bestNeighbor?.Coordinates;
    }

    public List<GridPoint>? BuildPathFrom(GridPoint startLocation, bool refresh = true)
    {
        if (refresh)
        {
            Refresh();
        }

        var startTile = GetTile(startLocation.ToString());
        if (startTile is null || !_covered[startTile.Id])
        {
            return null;
        }

        var startValue = _values[startTile.Id];
        if (startValue == int.MaxValue)
        {
            return null;
        }

        var path = new List<GridPoint> { startLocation };
        var current = startLocation;
        var currentValue = startValue;
        var timeCount = 0;

        while (currentValue > 0 && timeCount < 7850)
        {
            var next = GetNextStep(current, false);
            if (next is null)
            {
                return null;
            }

            path.Add(next.Value);
            current = next.Value;
            var currentTile = GetTile(current.ToString());
            if (currentTile is null)
            {
                return null;
            }

            currentValue = _values[currentTile.Id];
            timeCount++;
        }

        return currentValue == 0 ? path : null;
    }
}
